### PR TITLE
Ajaxify en form

### PR DESCRIPTION
**Pull request by comzeradd**
This seems to partially work. The EN API doesn't work properly locally here, so it's hard to test it properly, but if I get the render outside the if clauses it works fine.

@kirdia @pablocubico can you test it locally? it's my first attempt to ajaxify wordpress requests :D

enform

### CODE REVIEW

**Comment by pablocubico** [comment]
I'm getting a "0" response in place of the form. The response is `200` but it looks like it's just a "0" string. The typical response of a failed `admin-ajax.php` request is a `400` HTTP code with a single `0` string, but as this is a `200` I'm not sure what's going on. 🤔 

image

**Comment by kirdia** [comment]
I'm getting the same response as Pablo.
'0' content with 200 status code

File: classes/controller/blocks/class-enform-controller.php [hunk 1]
```diff
@@ -307,18 +331,18 @@ public function get_supporter_fields() : array {
...
+				$values                = $_POST['values'];
+
+				// CSRF protection.
+				$nonce                 = $_POST['_wpnonce'];
 
-				if ( ! wp_verify_nonce( $nonce, $nonce_action ) ) {
+				if ( ! wp_verify_nonce( $nonce, 'enform_submit' ) ) {
 					$data['error_msg'] = __( 'Invalid nonce!', 'planet4-engagingnetworks' );
-					return false;
+					return $data['error_msg'];
```
Comment (kirdia) [inline classes/controller/blocks/class-enform-controller.php]: you have to echo some output, this won't make it to the response with current implementation

File: classes/controller/blocks/class-enform-controller.php [hunk 2]
```diff
@@ -307,18 +331,18 @@ public function get_supporter_fields() : array {
...
+			// If this is an ajax call.
+			if ( wp_doing_ajax() ) {
 				$data['error_msg']     = '';
+				$values                = $_POST['values'];
+
+				// CSRF protection.
+				$nonce                 = $_POST['_wpnonce'];
 
-				if ( ! wp_verify_nonce( $nonce, $nonce_action ) ) {
+				if ( ! wp_verify_nonce( $nonce, 'enform_submit' ) ) {
```
Comment (kirdia) [inline classes/controller/blocks/class-enform-controller.php]: You are checking for the wrong nonce here.
I think wp creates and passes to frontend _wpnonce only in admin.
Generally the process is that you create a nonce (maybe for a specific ajax action), pass it to the frontend and then use it in the request.
I see an 'enform_nonce' passed in the ajax request, but I did not find where it is created.

File: classes/class-loader.php [hunk 3]
```diff
@@ -157,7 +157,15 @@ public function enqueue_public_assets() {
 			// Add master theme's main css as dependency for blocks css.
 			wp_enqueue_style( 'plugin-en', plugins_url( P4EN_PLUGIN_DIRNAME ) . '/style.css', [ ], $css_enform_creation );
 			// Add master theme's main js as dependency for blocks js.
-			wp_enqueue_script( 'plugin-en', plugins_url( P4EN_PLUGIN_DIRNAME ) . '/main.js', [ 'jquery' ], $js_enform_creation, true );
+			wp_register_script( 'plugin-en', plugins_url( P4EN_PLUGIN_DIRNAME ) . '/main.js', [ 'jquery' ], $js_enform_creation, true );
+			wp_localize_script(
+				'plugin-en',
+				'en_vars',
+				[
+					'ajaxurl' => admin_url( 'admin-ajax.php' ),
```
Comment (kirdia) [inline classes/class-loader.php]: Here you can pass the nonce created (or create a new one for the ajax action) and pass it to the frontend

**Comment by comzeradd** [comment]
I fixed the `nonce` problem. It was a misconfiguration on the twig template, on the `nonce_action` line.

**Comment by comzeradd** [comment]
Thanks @kirdia. It should work now. I still need to fix the error returns, but for valid data the form should now work.

**Comment by comzeradd** [comment]
I fixed the validation errors, so these now return on the template. We will need some design guidance on how to properly display errors, but I think it's good as a first iteration to showcase the form.

File: classes/controller/blocks/class-enform-controller.php [hunk 4]
```diff
@@ -307,18 +331,21 @@ public function get_supporter_fields() : array {
...
 		 * @return bool True if validation is ok, false if validation fails or if it is not a POST.
 		 */
-		public function handle_submit( &$data ) : bool {
-			if ( 'POST' === $_SERVER['REQUEST_METHOD'] ) {
-				// CSRF protection.
-				$nonce_action          = 'enform_submit';
-				$nonce                 = wp_create_nonce( $nonce_action );
-				$data['nonce_action']  = $nonce_action;
-				$data['enform_submit'] = 0;
+		public function handle_submit() {
```
Comment (angtheod) [inline classes/controller/blocks/class-enform-controller.php]: Method call in line 283 should be removed, since form submission was made asynchronous.

**Pull request merged**